ostd/src/mm/page_table/node/child.rs

```diff
@@ -0,0 +1,149 @@
+                    unsafe {
+                        Page::<PageTablePageMeta<E, C>>::inc_ref_count(paddr);
+                    }
+                }
+                // SAFETY: It will restore the node from the PTE in two cases,
+                // the first case is restoring a node that accounts for the
+                // increment of the reference count. The second case is
+                // restoring the PTE that is restored to a node. In both cases,
+                // the reference count matches the ownership.
+                Child::PageTable(unsafe { RawPageTableNode::from_paddr(paddr) })
+            } else {
+                match $is_tracked {
+                    MapTrackingStatus::Tracked => {
+                        if $inc_ref_cnt {
+                            // SAFETY: The PTE already holds a reference to the page.
+                            unsafe {
+                                DynPage::inc_ref_count(paddr);
+                            }
+                        }
+                        // SAFETY: Same as above. In both cases, the reference count
+                        // matches the ownership.
+                        Child::Page(unsafe { DynPage::from_raw(paddr) }, $pte.prop())
+                    }
+                    MapTrackingStatus::Untracked => Child::Untracked(paddr, $level, $pte.prop()),
+                    MapTrackingStatus::NotApplicable => panic!("Invalid tracking status"),
+                }
+            }
+        }
+    }};
+}
+
+impl<E: PageTableEntryTrait, C: PagingConstsTrait> Child<E, C>
+where
+    [(); C::NR_LEVELS as usize]:,
+{
+    /// Returns whether the child does not map to anything.
+    pub(in crate::mm) fn is_none(&self) -> bool {
+        matches!(self, Child::None)
+    }
+
+    /// Converts a child into a owning PTE.
+    ///
+    /// By conversion it loses information about whether the page is tracked
+    /// or not. Also it loses the level information. However, the returned PTE
+    /// takes the ownership (reference count) of the child.
+    ///
+    /// Usually this is for recording the PTE into a page table node. When the
+    /// child is needed again by reading the PTE of a page table node, extra
+    /// information should be provided using the [`Child::from_pte`] method.
+    pub(super) fn into_pte(self) -> E {
+        match self {
+            Child::PageTable(pt) => {
+                let pt = ManuallyDrop::new(pt);
+                E::new_pt(pt.paddr())
+            }
+            Child::Page(page, prop) => {
+                let level = page.level();
+                E::new_page(page.into_raw(), level, prop)
+            }
+            Child::Untracked(pa, level, prop) => E::new_page(pa, level, prop),
+            Child::None => E::new_absent(),
+        }
+    }
+
+    /// Converts a PTE back to a child.
+    ///
+    /// # Safety
+    ///
+    /// The provided PTE must be casted using [`Child::into_pte`]. And the
```

Comment:
Does this sentence means that the input pte have to be obtained from `Child::into_pte`? There are some APIs like `read_pte`, which will get a stored pte instead of constructing a new one. Personally, I think this sentence a bit ambiguous, as it might give the first impression that a pte obtained from `read_pte` cannot be passed into this function. I think it may need to add some additional clarifications, such as: The input pte must be a valid pte. A valid pte is one constructed by `Child::into_pte` and will be stored in the page table.

Reply:
There's absolutely no reason to cast a PTE that does not come from `Child::into_pte`. When you put it into a page table and read it back it still comes from `Child::into_pte` (although the hardware will modify some bits). I will polish the doc.